rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,376 @@
+query, and some protocol-specific (for example, PostgreSQL prepare statement)
+events. However, these recordings do not include the server's responses, which
+limits their usefulness for auditing and troubleshooting. By capturing the
+queries/commands and their responses, we can provide a more comprehensive view
+of database interactions. This enhancement will help organizations improve
+security audits, and facilitate debugging and operational reviews.
+
+## Details
+
+The database access recording is already in place for every database protocol,
+and is done in the same way as SSH recordings. The main difference is that all
+events are also emitted as regular audit events, which means they're accessible
+outside the recording files.
+
+The start/end and query events are recorded. In addition to those events,
+there are also protocol-specific events, for example `PostgresExecute` is
+emitted on PostgreSQL when a prepared statement is executed.
+
+The session recordings are available through the `tsh play` command, but they
+can only be viewed as JSON:
+
+
+```json
+[
+  {
+      "ei": 0,
+      "event": "db.session.start",
+      "code": "TDB00I",
+      "success": true,
+      ...
+  },
+  {
+      "ei": 1,
+      "event": "db.session.postgres.statements.execute",
+      "code": "TPG02I",
+      "portal_name": "test",
+      ...
+  },
+  {
+      "ei": 2,
+      "event": "db.session.query",
+      "code": "TDB02I",
+      "db_query": "SELECT 1;",
+      "success": true,
+      ...
+  },
+  {
+      "ei": 3,
+      "event": "db.session.end",
+      "code": "TDB01I",
+      ...
+  }]
+```
+
+### Audit events
+
+To have a complete recording with users input and server response, we'll
+include the command result event:
+
+```proto
+// DatabaseFieldType enum for database data fields data types.
+enum DatabaseDataFieldType {
+  UNSPECIFIED = 0;
+  STRING = 1;
+  NUMBER = 2;
+  DATE = 3;
+  BINARY = 4;
+}
+
+// DatabaseDataFieldMetadata contains metadata for database data fields.
+message DatabaseDataFieldMetadata {
+  // Name is the field name.
+  string Name = 1;
+  // Type is the field type. Since the values are stored inside the event as
+  // strings, this type can convert the row values back into their original
+  // type. This information can be used to format the row values.
+  DatabaseDataFieldType Type = 2;
+}
+
+// DatabaseDataRow represents a single data row, containing a list field values.
+message DatabaseDataRow {
+  // Values the row field values.
+  // Value is the row field values. The values aren not expected to be the
+  // direct database type but their string representation, making it possible to
+  // display them without parsing. For example, binary data can be encoded into
+  // base64.
+  repeated string Values = 1;
+}
+
+// DatabaseResultData database results data, containing fields metadata and also
+// the data itself.
+message DatabaseResultData {
```

Comment:
could you add a section on how database protocol is captured and converted to this? it can use PostgreSQL as example.